Update to HIA2 Reports Module

jasonrogena opened the pull request:
Adds the following to the HIA2 reports module:

 - Sorted indicators in the reports summary activity #518
 - Support for sending reports to opensrp-server #519 
 - Empty state illustration for the draft reports fragment #519 
 - Button widget in the native forms #520 
 - Limiting the number of reports fetched to the last three months for daily, and draft monthly reports #519, #517

jasonrogena committed:
JR| SQLite DB Overload Crash

Fix issue causing the database to crash when fetching daily and monthly
tallies

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
JR| Limiting HIA2 Reports

Limit monthly HIA2 reports to the last three months

Part of fix for issue #519

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
JR| Draft Monthly Form Values

Get database values, and add those to the HIA2 monthly reports form

Fix for issue #519

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
JR| Saving Draft Reports

Fix issue causing draft HIA2 reports not to be saved

Part of fix for issue #519

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
JR| Button Widget

Add a Button widget to the native forms, with initial support for next
step and end form behaviours

Part of fix for issue #520

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
JR| Clean up strings.xml

Clean up the strings.xml file in the opensrp-path module

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

Commit:
Update: added report empty state illustration #519

Signed-off-by: Ephraim Muhia <emuhia@ona.io>

jasonrogena committed:
JR| Sending HIA2 Reports

Hookup logic for sending monthly reports to the OpenSRP API

Part of fix for issue #519

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
Merge branch 'issue519' of github.com:OpenSRP/opensrp-client into issue519

jasonrogena committed:
JR| HIA2 Report Sync

Fix issues causing HIA2 reports not to be synced from the app

Part of fix for issue #519

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
JR| Sorting Indicators in ReportSummaryActivity

Fix the indicator sorting in the ReportSummaryActivity

Part of fix for issue #522

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
JR| Format Path values/dimens.xml

Format the main dimens file in the path module

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
JR| Code Formatting

Fix code formatting in JsonFormConstants, SentMonthlyFragment, and
sent_monthly_fragment.xml

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
JR| Raw Queries in MonthlyTalliesRepository

Remove all raw queries in MonthlyTalliesRepository

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

pld commented on android-json-form-wizard/src/main/java/com/vijay/jsonwizard/fragments/JsonFormFragment.java:
```diff
@@ -107,13 +107,33 @@ public boolean onOptionsItemSelected(MenuItem item) {
-            presenter.onNextClick(mMainView);
-            return true;
+            return next();
         } else if (item.getItemId() == R.id.action_save) {
```
this `if` looks like a `switch(item.getItemId())`

jasonrogena replied:
Cool. Android does not support using Android resource IDs (`R.id.action_next` for instance) in switch statements.

pld replied:
Ah, boo, OK

pld commented on android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/ButtonFactory.java:
```diff
@@ -0,0 +1,123 @@
+        button.setTag(R.id.address, stepName + ":" + jsonObject.getString("key"));
+
+        if (jsonObject.has("read_only")) {
+            button.setEnabled(!jsonObject.getBoolean("read_only"));
```
DRY up w/a local var?

pld commented on android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/ButtonFactory.java:
```diff
@@ -0,0 +1,123 @@
+                                    break;
+                            }
+                        }
+                    } catch (Exception e) {
```
this throws a generic exception? Is there anything more specific to catch instead?

pld commented on android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/ButtonFactory.java:
```diff
@@ -0,0 +1,123 @@
+                            String[] address = addressString.split(":");
+                            JSONObject jsonObject = ((JsonApi) context)
+                                    .getObjectUsingAddress(address);
+                            jsonObject.getJSONObject("action").put("result", false);
```
is there a reason to set this here and not after the switch, is something happening before the var is updated again in the switch

If not,

```
bool actionResult = false;
...
    case BEH...:
         actionResult = true;
...
jsonObject.getJSONObject("action").put("result", actionResult);
```
 
declare var here, replace lines 95 and 100 with var assignment, move this line to after switch and call with var

pld replied:
looking closer, you could make that var more generic `bool finishedFragment` (maybe?) move line 92 to after the switch and use the same var, remove lines 96 and 101

pld commented on opensrp-path/src/main/java/org/ei/opensrp/path/activity/HIA2ReportsActivity.java:
```diff
@@ -286,18 +200,34 @@ public void startMonthlyReportForm(String formName, Date date) {
+                    if (result.containsKey(FORM_KEY_CONFIRM)) {
+                        try {
+                            saveClicked = Boolean.valueOf(result.get(FORM_KEY_CONFIRM));
+                        } catch (Exception e) {
```
anything more specific?

jasonrogena replied:
I was going with handling all exceptions regardless the same way (in anticipation). To prevent crashes in the background bubbling up to the UI causing the app to stop.

pld commented on opensrp-path/src/main/java/org/ei/opensrp/path/repository/DailyTalliesRepository.java:
```diff
@@ -106,31 +107,51 @@ public void save(String day, Map<String, Object> hia2Report) {
-            database.endTransaction();
+            try {
+                database.endTransaction();
+            } catch (Exception e) {
```
anything more specific?

pld commented on opensrp-path/src/main/java/org/ei/opensrp/path/repository/DailyTalliesRepository.java:
```diff
@@ -146,18 +167,37 @@ public void save(String day, Map<String, Object> hia2Report) {
+
             cursor = getPathRepository().getReadableDatabase().query(TABLE_NAME, TABLE_COLUMNS,
-                    "strftime('%Y-%m', " + COLUMN_DAY + ") = '" + monthString+ "'",
+                    COLUMN_DAY + " >= " + startDate.getTimeInMillis() +
```
function taking `startDate` and `endDate` to create this string?

pld commented on opensrp-path/src/main/java/org/ei/opensrp/path/repository/DailyTalliesRepository.java:
```diff
@@ -192,19 +231,23 @@ public void save(String day, Map<String, Object> hia2Report) {
-                    .query(TABLE_NAME, TABLE_COLUMNS, null, null, null, null, null, null);
-            if (cursor != null && cursor.getCount() > 0 && cursor.moveToFirst()) {
-                while (!cursor.isAfterLast()) {
+                    .query(TABLE_NAME, TABLE_COLUMNS,
```
same about a function for this

pld commented on opensrp-path/src/main/java/org/ei/opensrp/path/repository/MonthlyTalliesRepository.java:
```diff
@@ -172,6 +186,32 @@ protected static void createTable(SQLiteDatabase database) {
+                    COLUMN_MONTH + " = '" + month + "'",
+                    null, null, null, null, null);
+            monthlyTallies = readAllDataElements(cursor);
+        } catch (Exception e) {
```
anything more specific?

pld commented on opensrp-path/src/main/java/org/ei/opensrp/path/repository/MonthlyTalliesRepository.java:
```diff
@@ -365,33 +396,47 @@ private MonthlyTally extractMonthlyTally(HashMap<Long, Hia2Indicator> indicatorM
-                        tally.setCreatedAt(new Date(dateStarted));
-                        tallies.add(tally);
-                }
+                    if (startDate != null) {
```
why not `&&` instead of two `if`s?

pld replied:
or better, two `&&`s and an `||` instead of four `if`s :)

pld reviewed (commented):
whenever exception was caught I wonder if there's a more specific exception, or maybe those are OK as catchalls? idk

cafootitt committed:
CG: Fix spacing issues in daily tally summary view

cagulas commented:
Lgtm

jasonrogena committed:
JR| Ignoring Indicators in Daily Tallies

Add functionality for ignoring some indicators in the daily tallies

Part of fix for issue #518

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
JR| Spacing in ReportSummaryActivity

Increase the column spacing in the ReportSummaryActivity

Part of fix for issue #518

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
JR| Code Formatting in ButtonFactory

Format code in the ButtonFactory class

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

pld commented:
Is there no limit to the kinds of exceptions that could be thrown by the
wrapped code?

jasonrogena committed:
JR| Exception Handling in ButtonFactory

Improve the exception handling in the ButtonFactory class

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
JR| Exception Handling in TallyRepositories

Improve the exception handling in DailyTalliesRepository, and
MonthlyTalliesRepository

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
JR| Cleanup in TalliesRepositories

Cleanup code in DailyTalliesRepository, and MonthlyTalliesRepository

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
JR| Code Cleaning in HIA2ReportsActivity

Cleanup code in HIA2ReportsActivity

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
JR| Dump HIA2 Indicators CSV in DB

Add code for redumping HIA2 indicators from the HAI2 indicators CSV file
to HIA2IndicatorsRepository

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

jasonrogena committed:
JR| getUniqueMonths in DailyTalliesRepository

Add the getUniqueMonths method in the DailyTalliesRepository

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>

keymane reviewed (approved):
- Need to test server side https://github.com/OpenSRP/opensrp-server/issues/260 
- Sending to DHIS2?

jasonrogena committed:
JR| CHN3-080 And CHN3-080-O HIA2 Indicators

Rename CHN3-80 to CHN3-080 and CHN3-80-O to CHN3-080-O as they were
changed in CSV file containing the indicators

Signed-off-by: Jason Rogena <jasonrogena@gmail.com>
